sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SyncRestProxy.java

```diff
@@ -0,0 +1,692 @@
+                methodParser.getReturnType(), context, options);
+        } catch (IOException e) {
+            throw LOGGER.logExceptionAsError(Exceptions.propagate(e));
+        }
+    }
+
+    @SuppressWarnings("deprecation")
+    void validateResumeOperationIsNotPresent(Method method) {
+        // Use the fully-qualified class name as javac will throw deprecation warnings on imports when the class is
+        // marked as deprecated.
+        if (method.isAnnotationPresent(com.azure.core.annotation.ResumeOperation.class)) {
+            throw LOGGER.logExceptionAsError(Exceptions.propagate(new Exception("'ResumeOperation' isn't supported.")));
+        }
+    }
+
+    static Context mergeRequestOptionsContext(Context context, RequestOptions options) {
+        if (options == null) {
+            return context;
+        }
+
+        Context optionsContext = options.getContext();
+        if (optionsContext != null && optionsContext != Context.NONE) {
+            context = CoreUtils.mergeContexts(context, optionsContext);
+        }
+
+        return context;
+    }
+
+    static Flux<ByteBuffer> validateLength(final HttpRequest request) {
+        final Flux<ByteBuffer> bbFlux = request.getBody();
+        if (bbFlux == null) {
+            return Flux.empty();
+        }
+
+        final long expectedLength = Long.parseLong(request.getHeaders().getValue("Content-Length"));
+
+        return Flux.defer(() -> {
+            final long[] currentTotalLength = new long[1];
+            return Flux.concat(bbFlux, Flux.just(VALIDATION_BUFFER)).handle((buffer, sink) -> {
+                if (buffer == null) {
+                    return;
+                }
+
+                if (buffer == VALIDATION_BUFFER) {
+                    if (expectedLength != currentTotalLength[0]) {
+                        sink.error(new UnexpectedLengthException(String.format(BODY_TOO_SMALL,
+                            currentTotalLength[0], expectedLength), currentTotalLength[0], expectedLength));
+                    } else {
+                        sink.complete();
+                    }
+                    return;
+                }
+
+                currentTotalLength[0] += buffer.remaining();
+                if (currentTotalLength[0] > expectedLength) {
+                    sink.error(new UnexpectedLengthException(String.format(BODY_TOO_LARGE,
+                        currentTotalLength[0], expectedLength), currentTotalLength[0], expectedLength));
+                    return;
+                }
+
+                sink.next(buffer);
+            });
+        });
+    }
+
+    static BinaryData validateLengthSync(final HttpRequest request) {
+        final BinaryData binaryData = request.getContent();
+        if (binaryData == null) {
+            return binaryData;
+        }
+
+        final long expectedLength = Long.parseLong(request.getHeaders().getValue("Content-Length"));
+
+        long length = binaryData.getLength();
```

Comment:
the `binaryData.getLength()` may return `null` which means it didn't measure length eagerly.
If this happens we should get the `BinaryDataContent` from inside and decorate it depending on type. I.e. if it's `FluxByteBufferContent` then fall back to the old `validateLength` or maybe just throw exception and don't accept Flux at all (I think this might be better). If it's `InputStreamContent` then decorate stream with `LengthValidatingInputStream`.

Reply:
I'd lean towards throwing when synchronous APIs are trying to send asynchronous payloads. And I think we're better off with having BinaryDataContent handling length validation instead of potentially having it crop up in many different spots in code (and likely having slightly skewed implementations, such as we've seen with InputStream -> Flux<ByteBuffer> conversions within the repo)